tf_quant_finance/black_scholes/vanilla_prices.py

```diff
@@ -182,6 +182,346 @@ def option_price(*,
+    asset_volatility = tf.convert_to_tensor(
+        asset_volatility, dtype=tf.float32, name="asset_volatility")
+    time_to_maturity = tf.convert_to_tensor(
+        time_to_maturity, dtype=tf.float32, name="time_to_maturity")
+    otype = tf.convert_to_tensor(
+        otype, dtype=tf.int32, name="otype")
+
+    # masks, Masks are used to sum appropriate integrals for approximation
+    down_and_in_call_tensor_lower_strike = tf.convert_to_tensor(
+        [1., 1., -1., -1., 0., 0., 1., 1., 1., 1., 0., 0.])
+    down_and_in_call_tensor_greater_strike = tf.convert_to_tensor(
+        [0., 0., 0., 0., 1., 1., 0., 0., 1., 1., 0., 0.])
+    down_and_in_put_tensor_lower_strike = tf.convert_to_tensor(
+        [1., 1., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0.])
+    down_and_in_put_tensor_greater_strike = tf.convert_to_tensor(
+        [0., 0., 1., 1., -1., -1., 1., 1., 0., 0., 1., 1.])
+    up_and_in_call_tensor_lower_strike = tf.convert_to_tensor(
+        [0., 0., 1., 1., -1., -1., 1., 1., 1., 1., 0., 0.])
+    up_and_in_call_tensor_greater_strike = tf.convert_to_tensor(
+        [1., 1., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0.])
+    up_and_in_put_tensor_lower_strike = tf.convert_to_tensor(
+        [0., 0., 0., 0., 1., 1., 0., 0., 1., 1., 0., 0.])
+    up_and_in_put_tensor_greater_strike = tf.convert_to_tensor(
+        [1., 1., -1., -1., 0., 0., 1., 1., 1., 1., 0., 0.])
+    down_and_out_call_tensor_lower_strike = tf.convert_to_tensor(
+        [0., 0., 1., 1., 0., 0., -1, -1., 0., 0., 1., 1.])
+    down_and_out_call_tensor_greater_strike = tf.convert_to_tensor(
+        [1., 1., 0., 0., -1., -1., 0., 0., 0., 0., 1., 1.])
+    down_and_out_put_tensor_lower_strike = tf.convert_to_tensor(
+        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1.])
+    down_and_out_put_tensor_greater_strike = tf.convert_to_tensor(
+        [1., 1., -1., -1., 1., 1., -1., -1., 0., 0., 1., 1.])
+    up_and_out_call_tensor_lower_strike = tf.convert_to_tensor(
+        [1., 1., -1., -1., 1., 1., -1., -1., 0., 0., 1., 1.])
+    up_and_out_call_tensor_greater_strike = tf.convert_to_tensor(
+        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1.])
+    up_and_out_put_tensor_lower_strike = tf.convert_to_tensor(
+        [1., 1., 0., 0., -1., -1., 0., 0., 0., 0., 1., 1.])
+    up_and_out_put_tensor_greater_strike = tf.convert_to_tensor(
+        [0., 0., 1., 1., 0., 0., -1., -1., 0., 0., 1., 1.])
+
+    # Constructing Masks
+    @tf.function
+    def get_out_map(params):
+      """
+      Function maps params to option pricing masks
+      Args:
+      params: Tuple of tensors. (barrier price, strike_price, option type)
+        params[0] = barrier_price
+        params[1] = strike_price
+        params[2] = otype
+      Returns:
+      returns mask used to price specified option
+      """
+      out_map = down_and_in_call_tensor_greater_strike # Default
+      if tf.math.less(params[1], params[0]):
+        if tf.equal(params[2], down_and_in_call):
```

Comment:
Note that in graph mode these comparisons won't work. However, if you use the previous suggestions, you won't need these comparisons.

Reply:
Modified to use tf.where!